### PR TITLE
Improves token refresh retries

### LINKED ISSUE #628: InvalidServicePrincipalCredentials on tags update (PUT)
Relevant e2e build is [here](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=30605457&view=logs&j=28b4d1e7-a465-5139-69ae-b80bf639468a&t=54e57b50-5801-5221-68d8-604b139a0916&l=186).

### PR DESCRIPTION
**Pull request by m1kola**
Fixes #628

### CODE REVIEW

File: pkg/util/aad/aad.go [hunk 1]
```diff
@@ -57,14 +61,13 @@ func GetToken(ctx context.Context, log *logrus.Entry, oc *api.OpenShiftCluster,
...
-			isAADSTS700016 := strings.Contains(err.Error(), "AADSTS700016")
+			log.Info(err)
 
 			// populate err with a user-facing error that will be visible if
 			// we're not successful.
-			log.Info(err)
 			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
 
-			if isAADSTS700016 {
+			if !autorest.IsTokenRefreshError(err) || strings.Contains(err.Error(), "AADSTS700016") {
```
Comment (jim-minter) [inline pkg/util/aad/aad.go:70]: this is not valid - you have overwritten err by this point.

File: pkg/util/aad/aad.go [hunk 2]
```diff
@@ -38,6 +39,9 @@ import (
 // tenant."`.  I think this can be returned when the service principal
 // associated with the application hasn't yet caught up with the application
 // itself.
+//
+// 3. Network failures.  If the error is not an adal.TokenRefreshError,
+// then it's likely a transient failure. For example, connection reset by peer.
```
Comment (jim-minter) [inline pkg/util/aad/aad.go:44]: I'm not sure I like this logical step at all.

File: pkg/util/aad/aad.go (hunk 1 above)
Comment (m1kola) [reply pkg/util/aad/aad.go:70]: Argh, right! I was goign to do a separate variable `isTokenRefreshError` initially for that reason. Tests will help to catch things like this, but it is going to be tricky to them it without refactoring. Will see what I can do when I have a moment for it.

**Comment by m1kola** [comment]
> I think the only way to do this is to be very specific about catching the connection reset by peer error and retrying it, and having unit test coverage.

It seems like catching only connection reset by peer is not enoguh. Here I'm trying to mimic what [SDK does](https://github.com/Azure/go-autorest/blob/e727cfcfc902307f3a34d6f6ce50748ca944786a/autorest/sender.go#L299-L304), but without `backoff`, `cap` and looking into `Retry-After` header (I think it is not important in this endpoint).

It looks like SDK re-treis on transient failure happened during deserialisation, as well. I think we should too. See [this](https://github.com/Azure/go-autorest/blob/e727cfcfc902307f3a34d6f6ce50748ca944786a/autorest/adal/token.go#L909-L912) and [this](https://github.com/Azure/go-autorest/blob/e727cfcfc902307f3a34d6f6ce50748ca944786a/autorest/adal/token.go#L924-L926).

> I think this is not the highest priority, and it's tricky. I am ok with abandoning or postponing for now if you want to - your call.

I think it's important. We caught it in e2e loop and I'm sure we will see this failure again and customers will see it. But I agree that it's not the highest priority. I'll be looking into it when I have a spare moment.

File: pkg/util/aad/aad.go [hunk 3]
```diff
@@ -58,13 +62,14 @@ func GetToken(ctx context.Context, log *logrus.Entry, oc *api.OpenShiftCluster,
...
 			isAADSTS700016 := strings.Contains(err.Error(), "AADSTS700016")
+			isTokenRefreshError := autorest.IsTokenRefreshError(err)
 
 			// populate err with a user-facing error that will be visible if
 			// we're not successful.
 			log.Info(err)
 			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
 
-			if isAADSTS700016 {
+			if !isTokenRefreshError || isAADSTS700016 {
```
Comment (jim-minter) [inline pkg/util/aad/aad.go:72]: I think this is worth a try, but we need some logging.

Comment (m1kola) [reply pkg/util/aad/aad.go:72]: @jim-minter there is `log.Info(err)` above which logs the original error.

Comment (jim-minter) [reply pkg/util/aad/aad.go:72]: good point, well made.

**Pull request merged**